src/lib/config/ModelConfig.cc

```diff
@@ -0,0 +1,997 @@
+namespace RISCVOpcode {
+#define GET_INSTRINFO_ENUM
+#include "RISCVGenInstrInfo.inc"
+}  // namespace RISCVOpcode
+
+/** AArch64 opcodes. Each opcode represents a unique AArch64 operation. */
+namespace AARCH64Opcode {
+#define GET_INSTRINFO_ENUM
+#include "AArch64GenInstrInfo.inc"
+}  // namespace AARCH64Opcode
+
+ModelConfig::ModelConfig(std::string path) {
+  // Reset ryml::Tree used to represent the config file
+  configTree_.clear();
+  configTree_.rootref() |= ryml::MAP;
+  isDefault_ = false;
+
+  std::ifstream file(path, std::ios::binary);
+  // Check for file existence
+  if (!file.is_open()) {
+    std::cerr << "[SimEng:ModelConfig] Could not read " << path << std::endl;
+    exit(1);
+  }
+  // Read in the contents of the file and create a ryml:Tree from it
+  std::stringstream buffer;
+  buffer << file.rdbuf();
+  configTree_ = ryml::parse_in_arena(ryml::to_csubstr(buffer.str()));
+  file.close();
+
+  // Set the expectations of the config file and validate the config values
+  // within the passed config file
+  setExpectations();
+  validate();
+}
+
+ModelConfig::ModelConfig() {
+  // Generate the default config file
+  generateDefault();
+}
+
+void ModelConfig::validate() {
+  missing_.clear();
+  invalid_.clear();
+
+  recursiveValidate(expectations_, configTree_.rootref());
+  postValidation();
+
+  std::string missingStr = missing_.str();
+  std::string invalidStr = invalid_.str();
+  // Print all missing fields
+  if (missingStr.length()) {
+    std::cerr << "[SimEng:ModelConfig] The following fields are missing from "
+                 "the provided "
+                 "configuration file:\n"
+              << missingStr << std::endl;
+  }
+  // Print all invalid values
+  if (invalidStr.length()) {
+    std::cerr << "[SimEng:ModelConfig] The following values are invalid for "
+                 "their associated field:\n"
+              << invalidStr << std::endl;
+  }
+  // Stop execution if the config file didn't pass checks
+  if (missingStr.length() || invalidStr.length()) exit(1);
+}
+
+void ModelConfig::reGenerateDefault(ISA isa, bool force) {
+  // Only re-generate the default config file if it hasn't already been
+  // generated for the specified ISA
+  if (!force && (ISA_ == isa && isDefault_)) return;
+  ISA_ = isa;
+  generateDefault();
+}
+
+void ModelConfig::generateDefault() {
+  // Reset ryml::Tree used to represent the config file
+  configTree_.clear();
+  configTree_.rootref() |= ryml::MAP;
+  isDefault_ = true;
+
+  // Set the expectations for the default config file, construct it, and
+  // validate it to ensure correctness for the simulation
+  setExpectations(true);
+  constructDefault(expectations_, configTree_.root_id());
+  validate();
+}
+
+void ModelConfig::constructDefault(ExpectationNode expectations,
+                                   size_t root_id) {
+  // Iterate over the expectations supplied
+  for (const auto& chld : expectations.getChildren()) {
```

Comment:
Personal preference for `child` over `chld` but not a required change to be made.

If change is made, please apply throughout file

Reply:
Changed